### ISSUE TITLE
License under an Apache license

### ISSUE THREAD

**Issue by glaforge** [javaparser/javaparser#144]
GPL and LGPL licensed libraries is often problematic when integrating in other projects.
Could it be possible to also license the project under the Apache license, or even MIT / BSD?

**Comment by ftomassetti** [comment]
Personally I am in favor of using the most liberal license but I think this would be a major decision and it would require every contributor to agree, but I will confess that I am completely ignorant on the possible legal requirements for changing license.
@matozoid @SmiddyPence @sebastiankirsch what do you think?

**Comment by glaforge** [comment]
IANAL, but you just need the major contributors to agree, so just a handful of persons being okay with that.

**Comment by ftomassetti** [comment]
@glaforge could you help us understand in which cases would help to have a different license? Do you have in mind anything specific?

**Comment by melix** [comment]
See http://www.apache.org/legal/resolved.html (LGPL section) : LGPL cannot be included in any Apache licensed product.

**Comment by glaforge** [comment]
As an example, if the Groovy project wanted to use this parser, it couldn't, because it's not allowed :-(

**Comment by ftomassetti** [comment]
Thank you guys, it helps, let's see what the other guys involved in the project think.

**Comment by glaforge** [comment]
Okie dokie.
The point is that if you want the project to be used widely, the current licenses are too restrictive :-(
The Groovy project might have a use for that parser, but unfortunately, with the license, we're not allowed to use it.

**Comment by matozoid** [comment]
I <3 BSD. I'm not making any money off this, and even if someone copies everything, erases my name and sells it, well, it's better than this stuff being used less because of the license.

**Comment by glaforge** [comment]
So that's a +1 vote :-)

**Comment by matozoid** [comment]
Aye, and there's a discussion on gitter going on about licenses. Not sure if you can see it, but we're wondering if you could explain what the problem with the license is.

**Comment by ftomassetti** [comment]
I agree 100% with @matozoid

**Comment by glaforge** [comment]
I went on gitter but didn't see any discussion about licenses.
Basically, the link given by @melix gives the point of view of the lawyers from the Apache Software Foundation, which explains that Apache licensed projects can't use GPL-ed / LGPL-ed code:
http://www.apache.org/legal/resolved.html#category-x
I'm definitely not a lawyer though, but if you want the project to be as open as possible, letting anybody use it however they see fit, etc, a more business-friendly license is needed.
GPL/LGPL are quite viral and would force ASL licensed project to use the same license, which complicate things, especially for projects hosted at the Apache Foundation.

**Comment by SmiddyPence** [comment]
+1

I'm pro freely available without limitations, i.e. whatever promotes widest use.  Also that there is no warranty offered.

From a migration to a new licence it would be nice to be blessed by @jgesser as the previous owner at code.google.  Maybe not not essential, we're in a major grey area given the long life of the project.

**Comment by ftomassetti** [comment]
+1 on getting @jgesser blessing

**Comment by ptitjes** [comment]
Hi all!

I'm not a big contributor yet, but you may have noticed I've got many stuff that I intend to push to JavaParser.

I'm glad that the Groovy project wants to use JavaParser, even if I don't really see the use case. (I think that would be cool if maybe you could describe that.)

I believe dual-licensing with ASL or BSD is not a move we can make without any thinking and understanding of the implications. I say dual-licensing because I want JavaParser to keep its free software licence.

I personally don't agree with the pressure put by businesses on free-software. I don't like either how that "open-source" wording got over the free-software spirit. I only do use GNU/Linux and love it.

So the question is what are the dual-licensing possibilities that would be compatible with ASL and still make it possible to have JavaParser packaged in free distros ?

**Comment by ptitjes** [comment]
@glaforge The (start of) discussion on gitter was in the private room. We keep the public room to welcome people and do tech support.

**Comment by jgesser** [comment]
Hi folks.

Thanks for remembering me.

When I chose LGPL I didn't know much about licenses. And I still don't. I just wanted to keep this code free and used only by free software. So LGPL seemed to be a good choice at that time.

But as you have mentioned, GPL or the lesser version have some restrictions, and changing the license can make possible this project to be more widely used.

So if my vote counts, +1. Please go ahead.

**Comment by glaforge** [comment]
Regarding the Groovy use cases, there are actually two places where we need Java parsing:
- We have a "joint compiler" which is able to compile both Groovy and Java sources together. The Groovy compiler needs to be aware of the Java sources (that haven't yet been compiled and which may depend on other Groovy source files) and has to parse those Java classes to be able to resolve references (methods, classes, signatures, etc...)
- We also have a tool which parses some special Java class from our code base, to produce some nice API documentation. So we need to parse that Java class somehow.

For the first use case, we actually have an Antlr grammar and parser, but it's not been updated to Java 8.
For the second use case, we used the QDox library, but it's not very active nor maintained.

JavaParser could be a nice replacement, which would support Java 8 out of the box.
We'd have to adapt our code base to use it, etc, so still some work, but as it seems more actively developed / enhanced / supported, it might be a good choice for us.
Except that the license prevents us from using it.

Regarding your point about "free distros", ASL licensed projects are also distributed with them. So I don't think there's any limitation or anything like that.

More generally speaking, there's as much contribution to ASL software as to (L)GPL software IMHO, if that's what concerns you.

**Comment by glaforge** [comment]
I forgot the question about what's compatible with Apache licensed software, and again, the Apache's legal FAQ gives the answer: http://www.apache.org/legal/resolved.html#category-b
If JavaParser were ASL too, of course, there wouldn't be any issue.

**Comment by ptitjes** [comment]
@glaforge Thanks for your answer. My concern was more about keeping the ability to use JavaParser inside free-software (meaning packaged and run as part of a free-software app). Also, I wonder if there are any legal holes to dual-licence (whatever the licences are, and also in the specific case of LGPL and ASL).

**Comment by glaforge** [comment]
If it's dual licensed, I don't see a reason why you would not still be able to make it used by other free-software apps.

I'm not aware of any legal holes regarding dual-licensing.

And as usual: IANAL ;-) — I'm not a lawyer, and I wouldn't want to become one :-D

**Comment by ptitjes** [comment]
@matozoid Just for the record, I personally don't agree with the "sells it" in

> even if someone copies everything, erases my name and sells it, well, it's better than this stuff being used less because of the license

On the contrary, I do think this project gives a free alternative to people. I don't want businesses to make money using our work without telling where it comes from, where users can contribute to it, ...

**Comment by ptitjes** [comment]
@glaforge Ok. Thank you for your informations.

@all IMO, I think the thing to do is finding some projects that do dual-licence LGPL and ASL for some time, and get their feedback.

**Comment by rpau** [comment]
Hi everyone,

Javaparser has copied the Java 8 support from javalang (https://github.com/rpau/javalang). You could verify it at https://github.com/javaparser/javaparser/issues/27. Morover, the project starts from the original Java 5 grammar (https://code.google.com/p/javaparser/wiki/UsingThisParser). Both projects have also LGPL license. Is still valid to have dual license?

**Comment by ftomassetti** [comment]
@rpau the maintainer of the original project (@jgesser) voted in favor of adopting different licenses. About the code originating from javalang, is it your project, correct? In that case you could grant us the permissions to move to a dual license (if the final outcome of this discussions is the decision to move to a dual license). Otherwise we would have to remove those changes to be able to move to a different license, correct?

@ptitjes I think that you should not be afraid about the possibility of using JavaParser in GPL software, if we adopt a dual license. However if you are afraid of possible usages of JavaParser in commercial projects you are right and you should vote against this change then.

**Comment by rpau** [comment]
@ftomassetti Yes, you are right in both questions. However, even I give my grants, we also need to establish how to proceed after the license has changed.

**Comment by ptitjes** [comment]
@ftomassetti Well, I am not against the use of JavaParser in commercial projects if this use is not hidden (mention of use, pointers to source code) and if those commercial projects invest in JavaParser. What I don't want is non-free copying/rebranding of our code.

**Comment by SmiddyPence** [comment]
Just to clarify, I would argue that it wasn't "copied" from javalang, the two grammars were merged.  There were (and still are) pre 1.8 concepts implemented differently between both projects grammars.  Arguably I copied they Java 8 concepts from javalang on your offer though @rpau, which was most appreciated ;-) The commit history will bare out:

A) Copy the javalang grammar `java_1_8.jj`
B) Created a `wip_java_1_8.jj` as a copy of javaparser `java_1_7.jj`
C) Merged `java_1_8.jj` Java8 concepts into `wip_java_1_8.jj`
D) Deleted the copied `java_1_8.jj`
E) Renamed the `wip_java_1_8.jj` to `java_1_8.jj`

My understanding (with no legal background). In principal you can only dual licence if the licences are compatible - in this case they are not.  In the case of ASF specifically they state: 

> The ASF will not dual-license our software because such licenses make it impossible to determine the conditions under which we have agreed to collaborate on a collective product, and are thus contrary to the Apache spirit of open, collaborative development among individuals, industry, and nonprofit organizations.
> [Reference](http://www.apache.org/licenses/GPL-compatibility.html)

[These folks](https://github.com/pyca/cryptography/issues/1209) see to have gone through the same thing, also consulting a lawyer.  Implying all committers need to agree :-(.  I would suggest doing something similar for legitimacy sake.

Also please can we keep the conversational elements the public gitter channel

**Comment by ptitjes** [comment]
@rpau When you say "we also need to establish how to proceed after the license has changed", do you mean to continue to exchange code between the projects ? I was wondering how much our code-base diverged, if it is technically possible that you directly use JavaParser now that it is more supported, how much work it would be for you ? Also, if technically possible, maybe you have reasons to keep your parser separate from JavaParser.

**Comment by rpau** [comment]
@ptitjes  I have enriched the AST with features I need for walkmod (interfaces to apply a semantic merge and also "symbol" related interfaces for semantic analysis, javadoc parsing, etc..). In fact, I am continuously, changing parts and deploying versions and I need this freedom. That's the reason why I have a separated project and some of my users have a strong coupling with my lib (which has a different package too).

@SmiddyPence these  "concepts" are "code" from my grammar.

**Comment by ptitjes** [comment]
@rpau This is what I assumed. But wanted to here it from you. :)

**Comment by SmiddyPence** [comment]
Fine @rpau ownership seems important to you.  I'm not going to argue with someone who claims IP over grammar.

```
| < ARROW: "->" >
| < DOUBLECOLON : "::">
```

Once again thank you for the use of your code.

**Comment by rpau** [comment]
@SmiddyPence For me the important think is that my project is LGPL and javaparser has common parts that had an important effort for me. Moreover, I have been helping with the bugs/problems regarding with the grammar I solved. So, I think that it is lawful that I manifest my opinion about that question.

**Comment by ftomassetti** [comment]
@rpau we are thankful for your help and I think it helped improving the quality of the project. I also understand the importance of having JavaParser available _also_ under a LGPL license. My question is: would you be contrary to grant us the permissions to release JavaParser _also_ under another license, making possible to use JavaParser in both Apache and GPL licensed projects? I think it would be great to keep the relations between projects (and if possible make it even tighter). The more eyes are looking at the same project the sooner we find and fix problems.

**Comment by rpau** [comment]
@ftomassetti I am not contrary to grant you permissions, because I understand the benefits that the project will receive. However, I would appreciate that walkmod appears in some place of acknowledgements in the webpage of this project. On the other hand, any new features, if these come from javalang, must be previously approved, because for me it is very important that my work is LGPL.

**Comment by SmiddyPence** [comment]
My understanding again with the IANAL predicate.  You can only have dual licencing if the licences are compatible... Apache and LGPL are not.  Apache also sound as though they do not dual licence in any case.

The second hand lawyer advise [here](https://github.com/pyca/cryptography/issues/1209) is that all contributors need to agree.  Therefor we would likely not be able to proceed if objections were made.  This is completely individuals prerogative, and understandable given they contributed their efforts under a difference licence.

Arguably off the back of this we should add a contributor agreement somewhere also!

**Comment by ftomassetti** [comment]
thank you @rpau ! I think it is fair to list the contributors. Your name is listed among the contributors in the pom file (https://github.com/javaparser/javaparser/blob/master/pom.xml) but I understand it is not very visible. I would suggest to add a reference to walkmod in the license section, something among the lines of: "This project is available under multiple license, users are allowed to choose betweens XXX and LGPL. This is possible because contributors granted permissions to do so. Among them Raquel Pau from the Walkmod project" and we could maybe add a link to the walkmod project. Of course this is just my suggestion, we should discuss it with all the people involved in the project but if I understand correctly  you are not going to prevent us to move to a dual license, if the community agrees on this, right?

@SmiddyPence yes, alas we would need some contributor agreement for the future :(

About the dual licenses we could perhaps adopt MIT and LGPL which seems to be compatible.

**Comment by ftomassetti** [comment]
And MIT should be compatible with Apache

**Comment by rpau** [comment]
@ftomassetti Right, but my desire is walkmod gains visibility, thats why I suggest to have a link to walkmod in some visible place :)

**Comment by glaforge** [comment]
You can dual license with LGPL and Apache.
My company (which I believe has double checked that with a lawyer in the past) is doing that with its Open Source project: http://restlet.com/downloads/legal-terms/

**Comment by ptitjes** [comment]
@SmiddyPence The fact that Apache does not authorize dual-licences _for the projects hosted under the Apache umbrella_ is an orthogonal matter to the ASL and LGPL being incompatible (well, as I understand it, their are incompatible in some points and maybe not in both ways... and that is to be verified).

Also I still hope we can find an additional licence that is compatible with LGPL and makes it possible to use it in an Apache project. (As you can see on their legal questions faq, they find all sorts of ways to do it, sometimes in strange ways)

I repeat that, as @rpau, I want my code and any future code I will produce for the JavaParser project to remain LGPL.

**Comment by sebastiankirsch** [comment]
> ​@ftomassetti https://github.com/ftomassetti Well, I am not against the
> use of JavaParser in commercial projects if this use is not hidden (mention
> of use, pointers to source code) and if those commercial projects invest in
> JavaParser. What I don't want is non-free copying/rebranding of our code.

+1 to that!

Also, as I have a ton on my plate, sorry for being off the channel the last
months (and some more to come)... :/

​​
2015-05-20 10:31 GMT+02:00 ptitjes notifications@github.com:

**Comment by ftomassetti** [comment]
It is really cool to we have a lot of people stating their opinions. I think it could be helpful to summarize the different opinions. The following is my understanding of the situation and please correct me when I am wrong.
- we can use a dual license scheme: this has been done by other projects and Apache License 2.0 and LGPL.
  This means that both Apache licensed and GPL licensed projects will be able to use JavaParser.
  The Groovy project could potentially consider using JavaParser if we do so
- even if the user chooses to use the project under the terms of the Apache License, attribution to the original authors should be granted:
  
  `If a NOTICE text file is included as part of the distribution of the original work, then derivative works must include a readable copy of these notices within a NOTICE text file distributed as part of the derivative works, within the source form or documentation, or within a display generated by the derivative works (wherever such third-party notices normally appear).`
- the persons who expressed their opinions so far are:
  - **ftomassetti**: in favor of dual license
  - **matozoid**: in favor of any change which makes the project more used (I assume it includes the dual license scheme, correct?)
  - **smyddypence**: in favor of any change which makes the project more used (I assume it includes the dual license scheme, correct?)
  - **ptitjes**: I am not sure if he could accept the dual license scheme. His main concerns so far have been:
    - possibility to legally combine the licenses
    - necessity to report the name of the original projects should be covered by the NOTICE text). 
    
    Can you please confirm if you have other doubts or you do not agree with the current answers to your  doubts? Given you seems the person more contrary to this proposal I want to be 100% sure to not missrepresent your opinion
  - **jgesser** (previous maintainer of the project) seems in favor of changing license, but he did not explicitly mentioned dual license
  - **rpau**: she be ok if we add more visibility to walkmod, as the source/inspirator of the support for Java 1.8. We could add acknowledgment in the README of the project and list contributors in the NOTICE. It is important for her that the project keeps being available under LGPL and it should be the case if we go for the dual license
  - **sebastiankirsch**: also for him the problem is granting the attribution to the original project/contributors and it should be addressed in the NOTICE text

Am I missing some major problem?
Do you think we should ask for the opinions of other contributors? Among the persons not involved we have only contributors with 7 or less commits on github
Could you please correct me if I misrepresented your opinion?
So far it seems to me that the persons who are contrary, are contrary for specific reasons which seems possible to address. I would like to know if I am wrong on this point and if there is someone absolutely contrary to a dual license scheme involving Apache and LGPL. I think that until now the code has been available under LGPL and everyone who contributed under this license is entitled to prevent his/her code to be released _also_ under another license. If this is the case please state you rightful objection.

**Comment by matozoid** [comment]
I have no issues with dual licensing. I'm all for giving credit to whoever wants it. I'm all for sending a message to the github accounts of all contributors, asking them to respond here.

**Comment by matozoid** [comment]
This discussion seems to have halted - @ftomassetti , would you like to drag it to a conclusion?

**Comment by ftomassetti** [comment]
I wrote my personal understanding of the current situation on Gitter, I will report also here.

Ideally we should build consensus finding a compromise through discussion.
However unfortunately this is not always possible. I understand this topic has moral implications and everyone should be free to do his choice.

For me now it is crucial to understand the position of @sebastiankirsch : I think he was not in favor to move to the Apache License only, but I am not sure if he could find the dual license acceptable (which grants attribution and let both GPL and Apache licensed projects to use JavaParser). @sebastiankirsch  has done major contributions and I deeply respect his opinion so I think that if this is not acceptable for him we should desist. (Sorry for putting pressure on you @sebastiankirsch but the discussion grinted to an halt).

My current understanding (please correct me if I am wrong) is that @rpau could agree on the dual license, if we make it more evident her contribution. Attribution to the contributors (including her) will be guaranteed also by the Apache License. The position of @ptitjes seems less clear but he seems generally against it even if some of his concerns were addressed.

**Comment by rpau** [comment]
Yes @ftomassetti , you are right.

**Comment by ftomassetti** [comment]
thanks for confirming @rpau !

**Comment by ptitjes** [comment]
Sorry, was doing stuff! My points have been addressed by some reading:

First, about the _possibility to legally combine the licenses_, the weaker license is the one to consider in case of dual-licensing.

As for my concerns about the _non-free copying/rebranding of our code_, the weakest points of the ALv2 are from my point of view :
- When someone modifies an Apache licensed software and redistributes it, it is not necessary to release the modified version of the source code.
- It is not necessary to submit the changes made to the source code to upstream.

This kinda makes it a no-go for me! :( That is I won't block you from go the dual-license path, as I really don't have much contribution to the project, but I won't contribute more but to a LGPL-only fork. (Sorry I know that is harsh to say but I want to be honest.)

**Comment by glaforge** [comment]
So, what's the final decision, no Apache license?

2015-05-25 10:03 GMT+02:00 ptitjes notifications@github.com:

> Sorry, was doing stuff! My points have been addressed by some reading:
> 
> First, about the _possibility to legally combine the licenses_, the
> weaker license is the one to consider in case of dual-licensing.
> 
> As for my concerns about the _non-free copying/rebranding of our code_,
> the weakest points of the ALv2 are from my point of view :
> - When someone modifies an Apache licensed software and redistributes
>   it, it is not necessary to release the modified version of the source code.
> - It is not necessary to submit the changes made to the source code to
>   upstream.
> 
> This kinda makes it a no-go for me! :( That is I won't block you to go the
> dual-license way, as I really don't have much contribution to the project,
> but I won't contribute more or to a LGPL-only fork. (Sorry I know that is
> harsh to say but I want to be honest.)
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/javaparser/javaparser/issues/144#issuecomment-105160132
> .

## 

Guillaume Laforge
Groovy Project Manager
Product Ninja & Advocate at Restlet http://restlet.com

Blog: http://glaforge.appspot.com/
Social: @glaforge http://twitter.com/glaforge / Google+
https://plus.google.com/u/0/114130972232398734985/posts

**Comment by matozoid** [comment]
We're doing a final round now, it takes ages with people not being available and everything.

**Comment by glaforge** [comment]
Okiedokie, looking forward to a positive outcome!

**Comment by matozoid** [comment]
Okay, everybody was contacted. If there is no veto in this issue before the fifth of july, we'll add the license.

**Comment by matozoid** [comment]
Nobody vetoed, we will add an Apache license. Who wants to pick it up?

**Comment by SmiddyPence** [comment]
Just to clarify what is required:
- Delete old COPYING
- Delete COPYING.LESSER
- Add LICENSE stating that the library is licensed under both
- Add LICENSE.LGPL verbatim
- Add LICENSE.APACHE verbatim
- Update all file sources 
- Update README.md

Anything else?

**Comment by ftomassetti** [comment]
Why remove COPYING?

About the source files, how should we change the current text? Would this be ok?

```
/*
 * Copyright (C) 2007-2010 Júlio Vilmar Gesser.
 * Copyright (C) 2011, 2013-2015 The JavaParser Team.
 *
 * This file is part of JavaParser.
 * 
 * JavaParser can be used either under the terms of
 * a) the GNU Lesser General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 * b) the terms of the Apache License 
 *
 * You should have received a copy of both licenses in LICENCE.LGPL and
 * LICENCE.APACHE. Please refer to those files for details.
 *
 * JavaParser is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 */
```

**Comment by SmiddyPence** [comment]
I would say it is the modern way to use `LICENSE` rather than `COPYING`

**Comment by ftomassetti** [comment]
I created an initial pull request to get us started: please check my wording... I am not a lawyer and not even a native speaker... in the PR I will share a script to update the test of all source files.

**Comment by ftomassetti** [comment]
This should be done now. Should we close this one?

**Comment by SmiddyPence** [comment]
Dual licensing now available in 2.2.1
